File: BeijingBeijing_assets/scripts/cell/Hand.py
import KBEngine
from KBEDebug import *

class Hand( KBEngine.Entity ):

    def __init__(self):
        KBEngine.Entity.__init__(self)
        DEBUG_MSG("Hand createCell %i" % self.id)

        ####################################################
        if self.isDebug == 1:
            dataName = "DebugHand" + str(self.isLeft)
            KBEngine.globalData[dataName] = self
        ####################################################

    def handInputSync(self, callerEntityID, key, value):
        if key == 0:
            #DEBUG_MSG("Trigger val=%i left=%i"% (value, self.isLeft))
            self.triggerState = value
        elif key == 1:
            #DEBUG_MSG("grip val=%i left=%i"% (value, self.isLeft))
            self.gripState = value

        #####################################
        if self.isDebug == 0:
            dataName = "DebugHand" + str(self.isLeft)
            if KBEngine.globalData.has_key(dataName):
                debugHand = KBEngine.globalData[dataName]
                debugHand.handInputSync(callerEntityID,key,value)

    def handAxisSync(self, callerEntityID, key, value):
        if key == 2:
            self.thumbstickAxis = value
        #####################################
        if self.isDebug == 0:
            dataName = "DebugHand" + str(self.isLeft)
            if KBEngine.globalData.has_key(dataName):
                debugHand = KBEngine.globalData[dataName]
                debugHand.handAxisSync(callerEntityID,key,value)

    def handLaserStateSync(self, callerEntityID, laser):
        self.otherClients.onGetLaserState(laser)
        #####################################
        if self.isDebug == 0:
            dataName = "DebugHand" + str(self.isLeft)
            if KBEngine.globalData.has_key(dataName):
                debugHand = KBEngine.globalData[dataName]
                debugHand.handLaserStateSync(callerEntityID,laser)

    def handGrabStateSync(self, callerEntityID, grab):
        self.otherClients.onGetGrabState(grab)
        #####################################
        if self.isDebug == 0:
            dataName = "DebugHand" + str(self.isLeft)
            if KBEngine.globalData.has_key(dataName):
                debugHand = KBEngine.globalData[dataName]
                debugHand.handGrabStateSync(callerEntityID,grab)

    def updateHandLocation(self, callerEntityID, pos):
        self.position = pos
        #####################################
        if self.isDebug == 0:
            dataName = "DebugHand" + str(self.isLeft)
            if KBEngine.globalData.has_key(dataName):
                debugHand = KBEngine.globalData[dataName]
                debugHand.updateHandLocation(callerEntityID,pos)

    def updateHandDirection(self, callerEntityID, rot):
        self.direction = rot
        #####################################
        if self.isDebug == 0:
            dataName = "DebugHand" + str(self.isLeft)
            if KBEngine.globalData.has_key(dataName):
                debugHand = KBEngine.globalData[dataName]
                debugHand.updateHandDirection(callerEntityID,rot)

File: Source/BeijingBeijing/Network/KBE_Entity/Hand.cpp
#include "Hand.h"
#include "EntityBase.h"
#include "KBEngine.h"
#include "Network/NetworkComponent/NetworkHandComponent.h"

ENTITYDEF_CLASS_REGISTER( Hand , EntityBase )

Hand::Hand() : EntityBase()
{
}

Hand::~Hand()
{
}

void Hand::__init__()
{
}

void Hand::onDestroy()
{
}

void Hand::sendHandLocation( const FVector &pos )
{
	FVector kbPos;
	UE4Pos2KBPos( kbPos , pos );

	cellCall( "updateHandLocation" , kbPos );
}

void Hand::sendHandRotation( const FRotator &rot )
{
	FVector kbRot;
	UE4Dir2KBDir( kbRot , rot );

	cellCall( "updateHandDirection" , kbRot );
}

void Hand::sendInputToOther( uint8 key , uint8 val )
{
	cellCall( "handInputSync" , key , val );
}

void Hand::sendAxisToOther( uint8 key , float val )
{
	cellCall( "handAxisSync" , key , val );
}

void Hand::set_triggerState( uint8 trigger )
{
	if ( !unrealObject || !unrealObject->IsValidLowLevel() ) return;
	UNetworkHandComponent *handComp = Cast<UNetworkHandComponent>( unrealObject );
	handComp->onGetInputAction( 0 , this->triggerState );
}

void Hand::set_gripState( uint8 grip )
{
	if ( !unrealObject || !unrealObject->IsValidLowLevel() ) return;
	UNetworkHandComponent *handComp = Cast<UNetworkHandComponent>( unrealObject );
	handComp->onGetInputAction( 1 , this->gripState );
}

void Hand::set_thumbstickAxis( float axis )
{
	if ( !unrealObject || !unrealObject->IsValidLowLevel() ) return;
	UNetworkHandComponent *handComp = Cast<UNetworkHandComponent>( unrealObject );
	handComp->onGetInputAxis( 2 , this->thumbstickAxis );
}

void Hand::sendLaserStateToOther( uint8 laser )
{
	cellCall( "handLaserStateSync" , laser );
}

void Hand::onGetLaserState( uint8 laser )
{
	if ( !unrealObject || !unrealObject->IsValidLowLevel() ) return;
	UNetworkHandComponent *handComp = Cast<UNetworkHandComponent>( unrealObject );
	handComp->onGetLaserState( laser );
}

void Hand::sendGrabStateToOther( uint8 grab )
{
	cellCall( "handGrabStateSync" , grab );
}

void Hand::onGetGrabState( uint8 grab )
{
	if ( !unrealObject || !unrealObject->IsValidLowLevel() ) return;
	UNetworkHandComponent *handComp = Cast<UNetworkHandComponent>( unrealObject );
	handComp->onGetGrabState( grab );
}
